src/pocketmine/event/Event.php

```diff
@@ -67,4 +86,147 @@ public function setCancelled(bool $value = true){
+				// paused but timed out
+				($this->pauseTimeoutFunc)($this);
+				$this->pauseTimeout = $this->pauseTimeoutFunc = null;
+			}
+
+			if($this->pauseTimeout === null){
+				// continue() or pauseTimeoutFunc has been called,
+				next($this->asyncQueue);
+				$this->doAsyncLoop();
+			}
+		}
+
+		if($this->asyncComplete){
+			$this->onComplete();
+			return true;
+		}
+		return false;
+	}
+
+	private function doAsyncLoop() : void{
+		/** @var RegisteredListener $listener */
+		while(($listener = current($this->asyncQueue)) !== false){
+			// while has more listener and not paused
+			if($listener->getPlugin()->isEnabled()){
+				try{
+					$listener->callEvent($this);
+				}catch(\Throwable $e){
+					MainLogger::getLogger()->critical(
+						Server::getInstance()->getLanguage()->translateString("pocketmine.plugin.eventError", [
+							$this->getEventName(),
+							$listener->getPlugin()->getDescription()->getFullName(),
+							$e->getMessage(),
+							get_class($listener->getListener())
+						]));
+					MainLogger::getLogger()->logException($e);
+				}
+			}
+			if($this->pauseTimeout !== null){ // paused
+				break;
+			}
+			next($this->asyncQueue);
+		}
+		$this->asyncComplete = current($this->asyncQueue) === false;
+	}
+
+	private function onComplete() : void{
+		// $this->asyncComplete is true here. Do not try to callAsyncEvent() immediately.
+		($this->asyncCompleteFunc)($this);
+
+		$this->async = false;
+		$this->asyncComplete = false;
+		$this->asyncQueue = null;
+		$this->pauseTimeout = null;
+		$this->pauseTimeoutFunc = null;
+	}
+
+	public function pause(int $ticks = 200, ?callable $onTimeout = null) : void{
+		if(!$this->async){
+			throw new \InvalidStateException("Could not pause non-async event; only events called with PluginManager->callAsyncEvent() can be paused");
+		}
+		if($ticks > Event::MAX_PAUSE_TICKS){
+			throw new \OutOfRangeException("Events must not be paused for more than " . Event::MAX_PAUSE_TICKS . " ticks");
+		}
+		if($this->pauseTimeout !== null){
+			throw new \InvalidStateException("Event has already been paused");
+		}
+
+		$this->pauseTimeout = Server::getInstance()->getTick() + $ticks;
+		$this->pauseTimeoutFunc = $onTimeout ?? [$this, "onTimeout"];
+	}
+
+	final public function continue() : void{
+		if($this->pauseTimeout === null){
+			throw new \InvalidStateException("Cannot continue a non-paused event");
+		}
+		$this->pauseTimeout = $this->pauseTimeoutFunc = null;
+		// do not execute anything directly after this, to prevent recursion stack overflow or concurrency due to the plugin executing something after this call
+	}
+
+	public function onTimeout() : void{
+		MainLogger::getLogger()->error(
+			Server::getInstance()->getLanguage()->translateString("pocketmine.plugin.eventTimeout", [
+				$this->getEventName(),
+				current($this->asyncQueue)->getPlugin()->getDescription()->getFullName(),
+				get_class(current($this->asyncQueue)->getListener())
+			]));
```

Comment:
I don't like this. It makes the Event base circularly dependent on other things.